sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala

```diff
@@ -592,6 +597,12 @@ object StructType extends AbstractDataType {
       case (leftUdt: UserDefinedType[_], rightUdt: UserDefinedType[_])
         if leftUdt.userClass == rightUdt.userClass => leftUdt
 
+      case (leftType, rightUdt: UserDefinedType[_])
+        if leftType == rightUdt.sqlType => leftType
+
+      case (leftUdt: UserDefinedType[_], rightType)
+        if leftUdt.sqlType == rightType => rightType
```

Comment:
@Fokko, sorry for my late response. I doubt if we should allow this case.

Currently, `merge` only allows the same types but UDT <> UDT's SQL types are not the same types. I think it makes less sense to allow this case alone.

Also, this https://github.com/apache/spark/pull/26644/files#r350486670 looks weird. `jsonValue` seems it should have JSON-serialized value of its own type.

Reply:
Well, the `UserDefinedType` extends `DataType`, similar to `TimestampType`, `StringType`, and any other type. The thing is that a UserDefinedType can be compatible with any other type. For example, it is allowed to merge an int into a long. This is an explicit choice by the developer.

Reply:
> it is allowed to merge an int into a long.

But this `StructType.merge` does not allow such type merging. Given that, it looks weird to allow only UDT.

```scala
scala> import org.apache.spark.sql.types._
import org.apache.spark.sql.types._

scala> StructType.merge(LongType, IntegerType)
org.apache.spark.SparkException: Failed to merge incompatible data types bigint and int
  at org.apache.spark.sql.types.StructType$.merge(StructType.scala:600)
  ... 49 elided
```